pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java

```diff
@@ -177,4 +179,19 @@
      */
     @Experimental
     void setReportFile(String reportFilename);
+
+
+    @Experimental
+    void setSourceEncoding(String sourceEncoding);
+
+    @Experimental
+    String getSourceEncoding();
+
+    /**
+     * Retrieve slice of source code from the information contained in
+     * {@link RuleViolation}
+     *
+     * <p>See {@link AbstractRenderer#getSourceCode(RuleViolation)} for default impl</p>
+     */
+    SourceCode getSourceCode(RuleViolation violation);
```

Comment:
SourceCode is CPD-specific API and should not be used here. Also, adding a `String getCodeSnippet()` method to RuleViolation seems like a much more appropriate place. Even if we use SourceCode in the implementation there, it won't be exposed in the API.